pkg/redis/pagination.go

```diff
@@ -0,0 +1,65 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/pkg/rpcmetadata"
+)
+
+type paginationOptionsKeyType struct{}
+
+var paginationOptionsKey paginationOptionsKeyType
+
+type paginationOptions struct {
+	limit  int64
+	offset int64
+	total  *int64
+}
+
+// NewContextWithPagination instructs the store to paginate the results.
+func NewContextWithPagination(ctx context.Context, limit, page int64, total *int64) context.Context {
+	md := rpcmetadata.FromIncomingContext(ctx)
+	if limit == 0 && md.Limit != 0 {
+		limit = int64(md.Limit)
+	}
+	if page == 0 && md.Page != 0 {
+		page = int64(md.Page)
+	}
+	if page == 0 {
+		page = 1
+	}
+	return context.WithValue(ctx, paginationOptionsKey, paginationOptions{
+		limit:  limit,
+		offset: (page - 1) * limit,
+		total:  total,
+	})
+}
+
+// SetPaginationTotal sets the total number of results inside the paginated context, if it was not set already.
+func SetPaginationTotal(ctx context.Context, total int64) {
```

Comment:
Usually functions/methods in Go do not mutate the input `context.Context` and I'd argue that they're not supposed to. It's way cleaner to just return `context.Context`. Is there a reason we're mutating the context here?
Why not just return `ctx`, if `total` is set and return `NewContextWithPagination` otherwise?

Reply:
I think that these values are put in the context in a unary or stream handler, i.e. where you cannot return the context.

Then, in an outer handler, after running inner handlers, these values are retrieved from the context and put as headers.

Reply:
The context itself is not mutated - `*total` is a pointer to a `total` variable that the caller provided - the context itself is still immutable.

Reply:
If `ctx` contained pagination options with `total == 0`, this would print `false`:
```go
origCtx := deepcopy.Copy(ctx)
SetPaginationTotal(ctx, 42)
fmt.Println(reflect.DeepEqual(ctx, origCtx))
```
Hence, the argument is mutated.